### DISCUSSION TITLE
Generated file size bigger than expected, possibly due to embedded basePDF file?

### DISCUSSION THREAD

**Discussion by JulianVallee** [pdfme/pdfme#51] [Q&A]
Hello,

my generated PDFs have a way bigger file size than expected, with an increase from 274kb to 807kb. I got curious and found that the library embeds+redraws the basePDF pages onto a new empty PDF document and after some testing, I think this is what is blowing up my file size. What are the reasons to do this extra step, instead of directly working with the basePDF document? Is there maybe a way to programmatically remove the embedded files from the generated PDF?

**Files for the reproduction steps:**
- [basePdf.pdf](https://github.com/pdfme/pdfme/files/8238717/basePdf.pdf)
- [generated.pdf](https://github.com/pdfme/pdfme/files/8238718/generated.pdf)
- [generatedRemovedEmbeds.pdf](https://github.com/pdfme/pdfme/files/8238719/generatedRemovedEmbeds.pdf)

**Steps to reproduce**
- "Manually" load and save the basePDF using `pdf-lib` => File size roughly stays the same
- Generate a PDF using the library => File size increases from 274kb to 807kb
- [Remove all embedded files from the generated PDF](https://avepdf.com/remove-pdf-content) => File size decreases from 807kb to a solid 243kb, can not see any optical difference
- Open the generated PDF in a hex editor => Found the TransformationMatrix values (1, 0, 0, 1, 0, 0) used in the library where a huge chunk of data was added

**What can be ruled out**
- Embedding fonts or not => negligible difference, 22kb in my case
- Adding or not adding text => negligible difference, few kb
- Embed an image => file size increases as expected

PS: I'm still using labelmake, but it is no longer in development and both projects embed the basePDF the same way... so I hope it's appropriate to ask here!

**Comment by hand-dot** [comment]
Hi! @JulianVallee 
Thanks for the report.

The increased size may be the data set in [`getEmbeddedPagesAndEmbedPdfBoxes`](https://github.com/pdfme/pdfme/blob/4bf14087a10d184b491640cc81b0349baff21e83/packages/generator/src/helper.ts#L95).

The pdfme layouts in absolute coordinates starting from the TOP,LEFT of the page.
So if the bounding box is not specified with top,left as 0, it will be out of place in some case.

However, something could be wrong because the data is too large.

I will investigate ways to prevent the data from increasing.

**Comment by JulianVallee** [comment]
Thank you @hand-dot, this pretty much confirms what I thought was happening.

I'm not sure if I fully understand the reason for touching the bounding boxes.. do you do this to invert the vertical coordinates, so that pdfme coordinates start at the top left?

After having another look at pdf-lib I found that no other methods allow changing the bounding box, unfortunately, so we also can't use the `copyPages/addPage` API.

I will post here when I found out something new.
